### PR TITLE
Handle specific arraycopy case

### PR DESCRIPTION
**Pull request by sehirst**
The arraycopy codegen was not expecting a specific pattern
where the second aladd had a NULL address child and an lsub
integer child. Since the address child is NULL we use the
integer child as the base register, but the code for cleaning
up the folded lsub assumed the index register was used. A simple
fix which selects the appropriate register in the memref
handles this case.

Signed-off-by: Simon Hirst <shirst@ca.ibm.com>

### CODE REVIEW

**Comment by sehirst** [comment]
@joransiu can you review while Filip is unavailable?

**Comment by Leonardo2718** [comment]
@genie-omr build zlinux, zos

**Comment by sehirst** [comment]
Removing WIP as my testing looks good

File: compiler/z/codegen/OMRMemoryReference.cpp [hunk 1]
```diff
@@ -1298,6 +1298,8 @@ OMR::Z::MemoryReference::populateAddTree(TR::Node * subTree, TR::CodeGenerator *
 
    noteAllNodesWithRefCountNotOne(nodesBefore, subTree, comp);
 
+   // Note: The addressChild for the 2nd aladd of an arraycopy can be NULL.
+   // This implies that the integer child is to be used directly as the address.
```
Comment (joransiu) [inline compiler/z/codegen/OMRMemoryReference.cpp]: This comment seems a bit out of place, as it does not relate to the code here.  I think, in general, populateMemoryReference(..) will just try to set _baseRegister over _indexRegister.   Where a typical base address node is an aconst NULL, the usual index subtree will be evaluated to fill _baseRegister and _offset instead.

File: compiler/z/codegen/OMRMemoryReference.cpp [hunk 2]
```diff
@@ -1407,11 +1409,37 @@ OMR::Z::MemoryReference::populateAddTree(TR::Node * subTree, TR::CodeGenerator *
...
+       * We will fold the lsub into the aladd's memref so the lRegLoad becomes the _indexRegister
+       * and the lconst becomes the offset.
+       * For the second aladd, the NULL addressChild means we use the integerChild as the _baseRegister.
+       * This is needed to prevent Zero Address Detection events on z/OS which happen when the base register of an instruction is zero.
+       * When we fold the lsub the lRegLoad becomes the _baseRegister and the lconst becomes offset.
+       * The important point is that in this case the _indexRegister is NULL.
+       * Thus we need to take care to select the appropriate register in the code below.
+       */
+      
+      TR::Register *integerChildRegister = _indexRegister ? _indexRegister : _baseRegister;
```
Comment (joransiu) [inline compiler/z/codegen/OMRMemoryReference.cpp]: Minor nit.  Can we make `(_indexRegister != NULL) ? ....` explicit here to improve readability?

**Comment by andrewcraik** [comment]
@sehirst over to you - please respond to the review comments when you have time.

File: compiler/z/codegen/OMRMemoryReference.cpp (hunk 1 above)
Comment (sehirst) [reply compiler/z/codegen/OMRMemoryReference.cpp]: Agreed. Deleted comment.

File: compiler/z/codegen/OMRMemoryReference.cpp (hunk 2 above)
Comment (sehirst) [reply compiler/z/codegen/OMRMemoryReference.cpp]: Changed.

**Comment by sehirst** [comment]
I've made the changes requested by Joran, so it's ready to go now.

**Comment by Leonardo2718** [comment]
@joransiu Would you mind taking another look and approving if it looks good? Given that @fjeremic is unavailable to merge this at the moment, I can take ownership of this.

**Comment by Leonardo2718** [comment]
@genie-omr build all

**Pull request merged**